sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -226,86 +169,102 @@ AZ_INLINE AZ_NODISCARD az_result az_token_credential_update(
       }
     } else {
-      az_mut_span_memset(token_buf, '#');
+      az_mut_span_memset(token_buf, 'X');
       return token_append_result;
     }
   }
 
   return AZ_OK;
 }
 
-AZ_INLINE AZ_NODISCARD az_result az_token_credential_get_token(
-    az_client_secret_credential * const credential,
-    az_span const request_url) {
-  clock_t const expiration = credential->token_credential.expiration;
-  if (expiration > 0) {
-    clock_t const clk = clock();
-    if (clk > 0 && clk < expiration) {
-      return AZ_OK;
-    }
-  }
-
-  uint8_t response_buf[AZ_TOKEN_CREDENTIAL_RESPONSE_BUF_SIZE] = { 0 };
+AZ_INLINE AZ_NODISCARD az_result
+_az_client_secret_credential_renew_token_credential(az_client_secret_credential * const self) {
+  uint8_t response_buf[_az_CLIENT_SECRET_CREDENTIAL_RESPONSE_BUF_SIZE] = { 0 };
 
   az_http_response http_response = { 0 };
   AZ_RETURN_IF_FAILED(az_http_response_init(
       &http_response, az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(response_buf))));
 
-  az_result const credential_update_result
-      = az_token_credential_update(credential, request_url, &http_response);
+  az_result const result = _az_client_secret_credential_ms_oauth2_get_token(self, &http_response);
 
   az_mut_span_memset(http_response.builder.buffer, '#');
-  return credential_update_result;
+  return result;
+}
+
+AZ_INLINE AZ_NODISCARD az_result
+_az_client_secret_credential_ensure_token_credential(az_client_secret_credential * const self) {
+  clock_t const expiration = self->_token_credential._expiration;
+  if (expiration > 0) {
+    clock_t const clk = clock();
+    if (clk > 0 && clk < expiration) {
+      return AZ_OK;
+    }
+  }
+
+  return _az_client_secret_credential_renew_token_credential(self);
 }
 
-static AZ_NODISCARD az_result az_token_credential_add_token_header(
-    az_client_secret_credential * const credential,
+static AZ_NODISCARD az_result _az_client_secret_credential_credential_func(
+    az_client_secret_credential * const self,
     az_http_request_builder * const hrb) {
-  AZ_CONTRACT_ARG_NOT_NULL(credential);
+  AZ_CONTRACT_ARG_NOT_NULL(self);
   AZ_CONTRACT_ARG_NOT_NULL(hrb);
 
-  AZ_RETURN_IF_FAILED(az_token_credential_get_token(credential, az_mut_span_to_span(hrb->url)));
-  AZ_RETURN_IF_FAILED(az_http_request_builder_append_header(
-      hrb, AZ_STR("authorization"), az_mut_span_to_span(credential->token_credential.token)));
+  {
+    az_url url = { 0 };
+    AZ_RETURN_IF_FAILED(az_succeeded(az_url_parse(az_span_builder_result(&hrb->url_builder), &url)));
+    if (!az_span_eq_ascii_ignore_case(url.scheme, AZ_STR("https"))) {
+      return AZ_ERROR_ARG;
+    }
+  }
 
-  return AZ_OK;
+  AZ_RETURN_IF_FAILED(_az_client_secret_credential_ensure_token_credential(self));
+
+  return az_http_request_builder_append_header(
+      hrb,
+      AZ_STR("authorization"),
+      (az_span){
+          .begin = self->_token_credential._token_buf,
+          .size = self->_token_credential._token_size,
+      });
```

Comment:
can we use az_mut_span_to_span(...)  instead of this?

Reply:
Unfortunately we can't because what we have is buffer+size, and not the span (I think it is unnecessary to store an extra pointer to `begin`).